File: app/src/main/java/at/shockbytes/dante/network/amazon/AmazonBookDownloader.java
package at.shockbytes.dante.network.amazon;

import javax.inject.Inject;

import at.shockbytes.dante.network.BookDownloader;
import at.shockbytes.dante.util.books.BookSuggestion;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * @author Martin Macheiner
 *         Date: 13.02.2017.
 */

public class AmazonBookDownloader implements BookDownloader {

    private final String ACCESS_KEY = "TODO";
    private final String ASSOCIATE_TAG = "TODO";

    private AmazonItemLookupApi api;

    @Inject
    public AmazonBookDownloader(AmazonItemLookupApi api) {
        this.api = api;
    }

    @Override
    public Observable<BookSuggestion> downloadBookSuggestion(String isbn) {
        return api.downloadBookSuggestion("AWSECommerceService", "ItemLookup", "Large",
                "All", "ISBN", isbn, ACCESS_KEY, ASSOCIATE_TAG, createTimestamp(),
                createSignature())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io());
    }

    private String createTimestamp() {
        return null;
    }

    private String createSignature() {
        return null;
    }

}

File: app/src/main/java/at/shockbytes/dante/network/amazon/AmazonBookDownloader.kt
package at.shockbytes.dante.network.amazon

import at.shockbytes.dante.network.BookDownloader
import at.shockbytes.dante.book.BookSuggestion
import at.shockbytes.dante.util.scheduler.SchedulerFacade
import io.reactivex.Observable

/**
 * Author:  Martin Macheiner
 * Date:    13.02.2017
 */
class AmazonBookDownloader(
    private val api: AmazonItemLookupApi,
    private val schedulerFacade: SchedulerFacade
) : BookDownloader {

    private val accessKey = "TODO"
    private val associateTag = "TODO"

    override fun downloadBook(isbn: String): Observable<BookSuggestion> {
        return api
            .downloadBookSuggestion(
                "AWSECommerceService",
                "ItemLookup",
                "Large",
                "All",
                "ISBN",
                isbn,
                accessKey,
                associateTag,
                createTimestamp(),
                createSignature()
            )
            .observeOn(schedulerFacade.ui)
            .subscribeOn(schedulerFacade.io)
    }

    private fun createTimestamp(): String {
        return ""
    }

    private fun createSignature(): String {
        return ""
    }
}
